### PR TITLE
Add multiprocessing for fit generator

### PR DESCRIPTION
**Pull request by tdeboissiere**
Implement fit_generator with python multiprocessing (#2730).

Sorry for the spam, I closed my previous PR by accident...
Anyway, this implementation now passes all the checks on Travis, has improved documentation and unit tests.

### CODE REVIEW

File: keras/engine/training.py [hunk 1]
```diff
@@ -395,36 +391,54 @@ def standardize_weights(y, sample_weight=None, class_weight=None,
 
 
 def generator_queue(generator, max_q_size=10,
-                    wait_time=0.05, nb_worker=1):
-    '''Builds a threading queue out of a data generator.
+                    wait_time=0.05, nb_worker=1, maxproc=2):
```
Comment (fchollet) [inline keras/engine/training.py]: `maxproc` should be `nb_worker`, for consistency.

File: keras/engine/training.py [hunk 2]
```diff
@@ -1252,7 +1266,7 @@ def predict_on_batch(self, x):
     def fit_generator(self, generator, samples_per_epoch, nb_epoch,
                       verbose=1, callbacks=[],
                       validation_data=None, nb_val_samples=None,
-                      class_weight={}, max_q_size=10):
+                      class_weight={}, max_q_size=10, maxproc=2):
```
Comment (fchollet) [inline keras/engine/training.py]: Same

File: keras/models.py [hunk 3]
```diff
@@ -589,6 +589,9 @@ def fit_generator(self, generator, samples_per_epoch, nb_epoch,
                 The generator is expected to loop over its data
                 indefinitely. An epoch finishes when `samples_per_epoch`
                 samples have been seen by the model.
+                Note that because of multiprocessing, the generator can't depend
+                on non picklable objects (such as a lazy-evaluated array) since it wont
```
Comment (fchollet) [inline keras/models.py]: I don't think "lazy-evaluated array" is a standard term. Can you clarify?

File: tests/keras/test_multiprocessing.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+        model.fit_generator(myGenerator(),
+                            samples_per_epoch=320,
+                            nb_epoch=1,
+                            verbose=1,
+                            max_q_size=10,
+                            maxproc=2)
+        reached_end = True
+
+    except Exception as e:
```
Comment (fchollet) [inline tests/keras/test_multiprocessing.py]: I don't get why you need a catch-all except around the entire function?

**Comment by tdeboissiere** [comment]
1. Clarification for lazy-evaluated array:

From what I understand, non-picklable object can't be passed to children process easily. This includes lazy evaluated arrays

So for instance, if you define a generator this way:

    arr_data = np.random.randint(0,256, (1000, 3, 224, 224))
    arr_labels = np.random.randint(0, 2, 1000)
    np.savez("data_dummy", **{"data": arr_data, "labels": arr_labels})
    
    arr = np.load("data_dummy.npz") # arr is non picklable

    def myGenerator():

        while True:
            X = arr["data"]
            y = arr["labels"]
            yield (X, y)`

Then `arr` won't be passed correctly to all the children processes and the code will throw an error (typically the generator will return None).

However, if you load `arr` in memory before defining the generator or if you define your generator as:

    def myGenerator():

        while True:
            X = arr["data"]
            y = arr["labels"]
            yield (X, y)

you won't have this issue.

2. Catch-all: you're right, I don't need the catch-all, it was merely a convenience when debugging.

File: keras/engine/training.py [hunk 5]
```diff
@@ -395,36 +391,85 @@ def standardize_weights(y, sample_weight=None, class_weight=None,
...
-                    wait_time=0.05, nb_worker=1):
-    '''Builds a threading queue out of a data generator.
+                    wait_time=0.05, nb_worker=2):
+    '''Builds a multiprocessing queue out of a data generator.
     Used in `fit_generator`, `evaluate_generator`, `predict_generator`.
+    Note that multiprocessing can't easily pass non picklable objects to
+    children processes. This notably means you can't pass the same lazy evaluated data array
+    to several children processes.
+
+    examples :
```
Comment (fchollet) [inline keras/engine/training.py]: If providing code examples in a docstring, it should use the MarkDown syntax for code snippets (as used elsewhere in docstring code snippets in the codebase) and the code should follow PEP8 syntax.

Alternatively, you could simply remove the code example, remove the mention of "lazy evaluated arrays", and simply state that the data should be picklable.

**Comment by fchollet** [comment]
Overall I think this would break a lot of people's code. For instance it would immediately break the `cifar10_cnn.py` example we have in Keras. It may well break a large majority of all data generators that are currently in use.

I'll think about it some more.

**Comment by tdeboissiere** [comment]
Sure. In the meantime I could push an example of how to build a generator with multiprocessing (along the line of the gist I posted in the original issue) if you think that is relevant

**Comment by carlthome** [comment]
I'd like something like this. It's not straightforward how to multi-thread fit_generator, evaluate_generator and test_generator from outside without introducing an additional queue.

**Comment by fchollet** [comment]
I think we should consider making it possible to have both the threading and the processing implementation (essentially we just need to change the imports) and have an argument for switching from one to the other (such as `pickle_safe`, which would default to False for backwards compatibility). How this works should be clearly explained in the documentation.

That way we can benefit from this PR while avoiding breaking everyone's code.

**Comment by tboquet** [comment]
I'm using Celery and it forks processes on workers (it's possible to control this behavior though). When the model are small enough it's possible to have multiple models running on the same GPU. It would maybe break if multiprocessing is enabled by default. +1 to have both threading and multithreading :+1:  .

**Comment by carlthome** [comment]
Cool! I'm guessing it's pretty uncommon to train multiple models on the same GPU though (mostly due to VRAM issues)? Regardless, I would be pretty content with threads, so how about just reintroducing nb_worker with the generator queue?

**Comment by tdeboissiere** [comment]
I have added multiprocessing as an option with the keyword pickle_safe as mentioned above.

Note that I have reintroduced the deprecated flag when using nb_worker > 1 and pickle_safe = False.

**Comment by fchollet** [comment]
> Note that I have reintroduced the deprecated flag when using nb_worker > 1 and pickle_safe = False.

Sounds good, it's much better to have a unified API.

LGTM, merging. Thanks for the contribution!

**Comment by mdeff** [comment]
@tdeboissiere thanks for your contribution. Shouldn't we make the `pickle_safe` limitation clearer in the `fit_generator` documentation ? I think the example you give above would help a lot. Maybe also warn people that external libraries used for pre-processing may use non-pickable objects.

**Comment by mikalyoung** [comment]
As of today March 5, 2017, does the most recent keras include this maximized speed of fit_generator (ie. using multiprocessing)? My current version does not seem to have this, as my fit_generator() is still about 4x slower than fit().

**Comment by mikalyoung** [comment]
Keras 1.2.2 does have this multiprocessing included, but my fit_generator() is still about 4x slower than fit(). Can this be further sped up?

**Pull request merged**